File: Smart-Bus/Smart-Bus/Classes/Main/ZoneControl/Zones/Controller/SHAreaControlViewController.m
//
//  SHAreaControlViewController.m
//  Smart-Bus
//
//  Created by Mark Liu on 2017/12/13.
//  Copyright © 2017年 SmartHome. All rights reserved.
//

#import "SHAreaControlViewController.h"
#import "SHZoneAudioPlayControlViewController.h"
#import "SHAreaSettingViewController.h"
 
@interface SHAreaControlViewController ()

// MARK: - 区域与类型

/// 当前区域
@property (assign, nonatomic) SHZone *currentZone;

/// 所有的设置类型ID
@property (strong, nonatomic) NSArray *allDeviceTypeIDs;

/// 系统相关的名称
@property (strong, nonatomic) NSMutableArray *deviceTypeNames;

// MARK: - tarBar区域

///  tarBarScrollView （要设置tarBar的滚动范围）
@property (strong, nonatomic) UIScrollView *tabBarScrollView;

/// 上一次选中的按钮
@property (weak, nonatomic) SHZoneControlButton *preivousButton;

@end

@implementation SHAreaControlViewController


// MARK: - 初始化设置导航

/// 设置区域
- (void)setArea {
    
    // 无操作权限 或者没有设备模块
    if ([SHAuthorizationViewController isOperatorDisable] ||
        !self.selectedViewController) {
        
        return;
    }
    
    SHSystemDeviceType deviceType = 0;
    
    /// 列表类
    if ([self.selectedViewController isKindOfClass:[SHZoneDevicesViewController class]]) {
        
        SHZoneDevicesViewController *zoneDevicesViewController = self.selectedViewController;
        
        switch (zoneDevicesViewController.deviceType) {
            
            case SHSystemDeviceTypeHvac: {
                
                deviceType = SHSystemDeviceTypeHvac;
            }
                break;
                
            case SHSystemDeviceTypeAudio: {
                
                deviceType = SHSystemDeviceTypeAudio;
            }
                break;
                
            case SHSystemDeviceTypeFloorHeating: {
                
                deviceType = SHSystemDeviceTypeFloorHeating;
            }
                break;
                
            case SHSystemDeviceTypeNineInOne: {
                
                deviceType = SHSystemDeviceTypeNineInOne;
            }
                break;
                
            case SHSystemDeviceTypeTv: {
                
                deviceType = SHSystemDeviceTypeTv;
            }
                break;
                
            case SHSystemDeviceTypeDvd: {
                
                deviceType = SHSystemDeviceTypeDvd;
            }
                break;
                
            case SHSystemDeviceTypeSat: {
                
                deviceType = SHSystemDeviceTypeSat;
            }
                break;
                
            case SHSystemDeviceTypeAppletv: {
                
                deviceType = SHSystemDeviceTypeAppletv;
            }
                break;
                
            case SHSystemDeviceTypeProjector: {
                
                deviceType = SHSystemDeviceTypeProjector;
            }
                break;
                
            case SHSystemDeviceTypeDmx: {
                 
                SHZoneDmxGroupSettingViewController *dmxGroupSettingController = [[SHZoneDmxGroupSettingViewController alloc] init];
                
                dmxGroupSettingController.currentZone = self.currentZone;
                
                [self.navigationController pushViewController:dmxGroupSettingController animated:YES];
 
                return;
            }
                break;
                
            default:
                break;
        }
        
        // 灯泡列表
    } else if ([self.selectedViewController isKindOfClass:[SHZoneLightViewController class]]){
        
        deviceType = SHSystemDeviceTypeLight;
    
        // 窗帘
    } else if ([self.selectedViewController isKindOfClass:[SHZoneShadeViewController class]]){
        
        deviceType = SHSystemDeviceTypeShade;
    
        // Mood
    } else if ([self.selectedViewController isKindOfClass:[SHZoneMoodViewController class]]){ 
        
        SHZoneControlEditMoodViewController *editMoodController = [[SHZoneControlEditMoodViewController alloc] init];
        
        editMoodController.currentZone = self.currentZone;
        
        [self.navigationController pushViewController:editMoodController animated:YES];
        
        return; // Mood是不同的方式
    
        // Fan
    } else if ([self.selectedViewController isKindOfClass:[SHZoneFanViewController class]]){
     
        deviceType = SHSystemDeviceTypeFan;
     
        // 干节点
    } else if ([self.selectedViewController isKindOfClass:[SHZoneDryContactViewController class]]){

        deviceType = SHSystemDeviceTypeDryContact;
       
        // 温度传感器
    } else if ([self.selectedViewController isKindOfClass:[SHZoneTemperatureSensorViewController class]]){
        
        deviceType = SHSystemDeviceTypeTemperatureSensor;
    
        // scene控制
    } else if ([self.selectedViewController isKindOfClass:[SHSceneControlViewController class]]){
        
        deviceType = SHSystemDeviceTypeSceneControl;
        
        // Sequence 控制
    } else if ([self.selectedViewController isKindOfClass:[SHSequenceControlViewController class]]){
        
        deviceType = SHSystemDeviceTypeSequenceControl;
        
        // otherControl 控制
    } else if ([self.selectedViewController isKindOfClass:[SHOtherControlViewController class]]){
        
        deviceType = SHSystemDeviceTypeOtherControl;
    }
    
    // 进入设置页面
     SHSystemDetailViewController *systemDetailViewController = [[SHSystemDetailViewController alloc] init];
    
    systemDetailViewController.currentZone =
        [self.selectedViewController currentZone];
  
    systemDetailViewController.systemType = deviceType;
    
    [self.navigationController pushViewController:systemDetailViewController animated:YES];
}

/// 设置导航栏
- (void)setUpNavigationBar {
    
    self.navigationItem.title = self.currentZone.zoneName;
    
    self.navigationItem.rightBarButtonItem =
        [UIBarButtonItem barButtonItemWithImageName:@"setting"
                              hightlightedImageName:@"setting"
                                          addTarget:self
                                             action:@selector(setArea)
                                             isLeft:NO
        ];
    
   
//    self.navigationItem.rightBarButtonItem.customView.hidden = !self.childViewControllers.count;
    
    // 由结构的变化 Navi Push TabBar 所以在这里再写一次
    UIView *leftCustomView = self.navigationItem.leftBarButtonItem.customView;
    UIView *rightCustomView = self.navigationItem.rightBarButtonItem.customView;
    
    CGFloat itemSize =  [UIDevice is_iPad] ?
        (navigationBarHeight + statusBarHeight) : tabBarHeight;
    
    leftCustomView.bounds = CGRectMake(0, 0, itemSize, itemSize);
    rightCustomView.bounds = CGRectMake(0, 0, itemSize, itemSize);
    
    //    leftCustomView.backgroundColor = [UIColor orangeColor];
    //    rightCustomView.backgroundColor = [UIColor greenColor];
}


// MARK: - 初始化UI

- (void)viewWillAppear:(BOOL)animated {

    [super viewWillAppear:animated];
    
    [self setUpNavigationBar];
}


/// 初始化控制器时设置 Zone
- (instancetype)initWithZone:(SHZone *)zone {
    
    if (self = [super init]) {
        
        self.currentZone = zone;
        [self addChildViewControllers];
        [self setUpTabBar];
    }
    
    return self;
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    
}

// MARK: -  tabBar

- (void)setUpTabBar {

    self.tabBar.backgroundImage = [UIImage resizeImage:@"tabBarbackground"];
    
    [self.view addSubview:self.tabBarScrollView];
    
    for (int i = 0; i < self.childViewControllers.count; i++) {
        
        SHZoneControlButton *zoneControlButton = [[SHZoneControlButton alloc] init];
        
        zoneControlButton.tag = i;
        
        // 设置图片
        //  [zoneControlButton setBackgroundImage:[UIImage resizeImage:@"tabBarSelectedbackground"] forState:UIControlStateSelected];
        
        [zoneControlButton setImage:[UIImage imageNamed:[self.deviceTypeNames[i] stringByAppendingString:@"TabBar"]] forState:UIControlStateNormal];
        
        [zoneControlButton setImage:[UIImage imageNamed:[self.deviceTypeNames[i] stringByAppendingString:@"TabBar_highlighted"]] forState:UIControlStateSelected];
        
        // 获得按钮名称 -- 取消文字显示
        //        [zoneControlButton setTitle:[[SHLanguageTools shareLanguageTools] getTextFromPlist:@"MAIN_PAGE" withSubTitle:self.systemNames[i]] forState:UIControlStateNormal];
        
        // ========== 不同的部分
        [zoneControlButton addTarget:self action:@selector(changeViewController:) forControlEvents:UIControlEventTouchUpInside];
        
        [self.tabBarScrollView addSubview:zoneControlButton];
        
        if (!i) { // 默认第一个是选中状态
            
            [self changeViewController:zoneControlButton];
        }
    }
}

/// 修改状态
- (void)changeViewController:(SHZoneControlButton *)button {
    
    if (button == self.preivousButton ) {
        return;
    }
    
    button.selected = !button.selected;
    self.preivousButton.selected = !self.preivousButton.selected;
    self.preivousButton = button;
    
    if (button.selected) {
        
        [self setSelectedIndex:button.tag];
    }
}

// MARK: - 添加子控制器

/// 添加区域中的所有子控制器
- (void)addChildViewControllers {

    self.allDeviceTypeIDs =
        [SHSQLiteManager.shared
            getSystemIDs:self.currentZone.zoneID];
    
    if (!self.allDeviceTypeIDs.count) {
        
        [SVProgressHUD showInfoWithStatus:SHLanguageText.noData];
        
        return;
    }
    
    // 查询
    for (NSUInteger i = 0; i < self.allDeviceTypeIDs.count; i++) {
        
        NSUInteger systemIDValue = [self.allDeviceTypeIDs[i] integerValue];

        
        switch (systemIDValue) {
                
            case SHSystemDeviceTypeLight: {
                
                SHZoneLightViewController *zoneLight = [[SHZoneLightViewController alloc] init];
                zoneLight.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"alight"];
                
                [self addChildViewController:zoneLight];
            }
                break;
                
            case SHSystemDeviceTypeHvac: {
                
                SHZoneDevicesViewController *zoneHVAC = [[SHZoneDevicesViewController alloc] init];
                
                zoneHVAC.currentZone = self.currentZone;
                zoneHVAC.deviceType = SHSystemDeviceTypeHvac;
                
                [self.deviceTypeNames addObject:@"bhvac"];
                
                [self addChildViewController:zoneHVAC];
                
            }
                break;
                
            case SHSystemDeviceTypeAudio: {
 
                SHZoneDevicesViewController *zoneAudio = [[SHZoneDevicesViewController alloc] init];
                
                zoneAudio.currentZone = self.currentZone;
                zoneAudio.deviceType = SHSystemDeviceTypeAudio;
                
                [self.deviceTypeNames addObject:@"czaudio"];
                [self addChildViewController:zoneAudio];
     
            }
                break;
                
            case SHSystemDeviceTypeShade: {
                
                SHZoneShadeViewController *zoneShade = [[SHZoneShadeViewController alloc] init];
        
                zoneShade.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"dshades"];
                
                [self addChildViewController:zoneShade];
      
            }
                break;
                
            case SHSystemDeviceTypeTv: {
 
                SHZoneDevicesViewController *zoneTV = [[SHZoneDevicesViewController alloc] init];
                
                zoneTV.currentZone = self.currentZone;
                zoneTV.deviceType = SHSystemDeviceTypeTv;
                
                [self.deviceTypeNames addObject:@"etv"];
                
                [self addChildViewController:zoneTV];
            }
                break;
                
            case SHSystemDeviceTypeSat: {
                
                SHZoneDevicesViewController *zoneDVD = [[SHZoneDevicesViewController alloc] init];
                
                zoneDVD.currentZone = self.currentZone;
                zoneDVD.deviceType = SHSystemDeviceTypeSat;
                
                [self.deviceTypeNames addObject:@"esat"];
                
                [self addChildViewController:zoneDVD];
            }
                break;
                
            case SHSystemDeviceTypeDvd: {
                
                SHZoneDevicesViewController *zoneDVD = [[SHZoneDevicesViewController alloc] init];
                
                zoneDVD.currentZone = self.currentZone;
                zoneDVD.deviceType = SHSystemDeviceTypeDvd;
                
                [self.deviceTypeNames addObject:@"edvd"];
                
                [self addChildViewController:zoneDVD];
                
            }
                break;
                
            case SHSystemDeviceTypeAppletv: {
                
                SHZoneDevicesViewController *zoneDVD = [[SHZoneDevicesViewController alloc] init];
                
                zoneDVD.currentZone = self.currentZone;
                zoneDVD.deviceType = SHSystemDeviceTypeAppletv;
                
                [self.deviceTypeNames addObject:@"eappletv"];
                
                [self addChildViewController:zoneDVD];
            }
                break;
                
            case SHSystemDeviceTypeProjector: {
                
                SHZoneDevicesViewController *zoneProjector = [[SHZoneDevicesViewController alloc] init];
                
                zoneProjector.currentZone = self.currentZone;
                zoneProjector.deviceType = SHSystemDeviceTypeProjector;
                
                [self.deviceTypeNames addObject:@"eprojector"];
                
                [self addChildViewController:zoneProjector];
            }
                break;
                
            case SHSystemDeviceTypeMood: {
                
                SHZoneMoodViewController *zoneMood = [[SHZoneMoodViewController alloc] init];
              
                zoneMood.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"fmoods"];
                
                [self addChildViewController:zoneMood];
            }
                break;
                
            case SHSystemDeviceTypeFan: {
        
                SHZoneFanViewController *zoneFan = [[SHZoneFanViewController alloc] init];
                
                zoneFan.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"gfan"];
                
                [self addChildViewController:zoneFan];
              
            }
                break;
            
            case SHSystemDeviceTypeFloorHeating: {
                
                SHZoneDevicesViewController *floorHeating = [[SHZoneDevicesViewController alloc] init];
                
                floorHeating.currentZone = self.currentZone;
                floorHeating.deviceType = SHSystemDeviceTypeFloorHeating;
                
                [self.deviceTypeNames addObject:@"hFloorheating"];
                
                [self addChildViewController:floorHeating];
            }
                break;
              
            case SHSystemDeviceTypeNineInOne: {
                
                SHZoneDevicesViewController *nineInOne = [[SHZoneDevicesViewController alloc] init];
                
                nineInOne.currentZone = self.currentZone;
                nineInOne.deviceType = SHSystemDeviceTypeNineInOne;
                
                [self.deviceTypeNames addObject:@"jnieInOne"];
                
                [self addChildViewController:nineInOne];
            }
                break;
           
            case SHSystemDeviceTypeDryContact: {
                
                SHZoneDryContactViewController *dryContact = [[SHZoneDryContactViewController alloc] init];
                    
                dryContact.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"kdryInput"];
                
                [self addChildViewController:dryContact];
            }
                break;
                
            case SHSystemDeviceTypeTemperatureSensor: {
            
                SHZoneTemperatureSensorViewController *temperatureSensor = [[SHZoneTemperatureSensorViewController alloc] init];
                
                temperatureSensor.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"ltemperature"];
                
                [self addChildViewController:temperatureSensor];
            }
                break;
                
            case SHSystemDeviceTypeDmx: {
                
                SHZoneDevicesViewController *dmx = [[SHZoneDevicesViewController
                                                     alloc] init];
                
                dmx.currentZone = self.currentZone;
                dmx.deviceType = SHSystemDeviceTypeDmx;
                [self.deviceTypeNames addObject:@"mdmx"];
                
                [self addChildViewController:dmx];
            }
                break;
                
            case SHSystemDeviceTypeSceneControl: {
                
                SHSceneControlViewController *sceneController = [[SHSceneControlViewController alloc] init];
                
                sceneController.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"nscene"];
                
                [self addChildViewController:sceneController];
            }
                break;
                
            case SHSystemDeviceTypeSequenceControl: {
                
                SHSequenceControlViewController
                * sequenceController =
                    [[SHSequenceControlViewController alloc] init];
                
                sequenceController.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"psequence"];
                
                [self addChildViewController:sequenceController];
            }
                break;
                
            case SHSystemDeviceTypeOtherControl: {
               
                SHOtherControlViewController *otherController = [[SHOtherControlViewController alloc] init];
                
                otherController.currentZone = self.currentZone;
                
                [self.deviceTypeNames addObject:@"qother"];
                
                [self addChildViewController:otherController];
            }
                break;
                
            default:
                break;
        }
    }
}

// MARK: -  布局

/// 布局子控件
- (void)viewDidLayoutSubviews {

    [super viewDidLayoutSubviews];
    
    // 保证iPhoneX时的选项卡位置大小和系统一致
    self.tabBarScrollView.frame = CGRectMake(0, self.view.bounds.size.height - self.tabBar.frame_height, self.view.frame.size.width, tabBarHeight);
   
    CGFloat itemHeight = self.tabBarScrollView.frame_height;
    CGFloat itemWidth = self.view.frame_width / 4;
    
    CGFloat startX = (self.view.frame_width - (self.childViewControllers.count > 4 ? 4 : self.allDeviceTypeIDs.count) * itemWidth) * 0.5;
    
    for (NSUInteger i = 0; i < self.tabBarScrollView.subviews.count; i++) {
        
        SHZoneControlButton *zoneControllerButton = self.tabBarScrollView.subviews [i];
        
        zoneControllerButton.frame = CGRectMake((zoneControllerButton.tag) * itemWidth + startX, (self.tabBarScrollView.frame_height - itemHeight) * 0.5, itemWidth, itemHeight);
    }
    
    self.tabBarScrollView.contentSize = CGSizeMake(self.childViewControllers.count * itemWidth, itemHeight);
}

// MARK: - getter && setter

/// tabBarScrollView
- (UIScrollView *)tabBarScrollView {

    if (!_tabBarScrollView) {
        
        _tabBarScrollView = [[UIScrollView alloc] init];
        _tabBarScrollView.scrollEnabled = YES;
        _tabBarScrollView.showsVerticalScrollIndicator = NO;
        _tabBarScrollView.showsHorizontalScrollIndicator = NO;
        _tabBarScrollView.pagingEnabled = NO;
    }
    
    return _tabBarScrollView;
}

/// 设备图片名称
- (NSMutableArray *)deviceTypeNames {
    
    if (!_deviceTypeNames) {
        _deviceTypeNames = [NSMutableArray array];
    }
    return _deviceTypeNames;
}

@end

File: Smart-Bus/Smart-Bus/Classes/Main/ZoneControl/Zones/Controller/SHAreaControlViewController.swift
//
//  SHAreaControlViewController.swift
//  Smart-Bus
//
//  Created by Mark Liu on 2017/12/13.
//  Copyright © 2017年 SmartHome. All rights reserved.
//

import UIKit

class SHAreaControlViewController: SHTabBarController {
    
    /// 当前区域
    var currentZone: SHZone?
    
    /// 设备类型ID号
    private lazy var deviceTypeIDs = [UInt]()
    
    /// 设备类型图标名称
    private lazy var deviceTypeNames = [String]()
    
    /// 选项卡
    private lazy var tabBarScrollView: UIScrollView = {
        
        let scrollView = UIScrollView()
        
        scrollView.isScrollEnabled = true
        scrollView.showsVerticalScrollIndicator = false
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.isPagingEnabled = false
        
        return scrollView
    }()

    /// 上一次选中的按钮
    var preivousButton: SHZoneControlButton?
    
    /// 设置控器
    func setupViewController(_ zone: SHZone?) {
        
        currentZone = zone
        
        addChildViewControllers()
        setUpTabBar()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        setupNavigationBar()
    }
}


// MARK: - 导航栏设置
extension SHAreaControlViewController {
    
    /// 设置区域
    @objc private func setArea() {
        
        if SHAuthorizationViewController.isOperatorDisable() {
            
            return
        }
        
        guard let selectController = selectedViewController else {
            return
        }
        
        var deviceType = SHSystemDeviceType.undefined
        
        // 列表类
        if selectController.isKind(of: SHZoneDevicesViewController.self) {
         
            let devicesController =
                selectController as!
                SHZoneDevicesViewController
            
            switch devicesController.deviceType {
                
            case .hvac:
                deviceType = .hvac
                
            case .audio:
                deviceType = .audio
                
            case .floorHeating:
                deviceType = .floorHeating
                
            case .nineInOne:
                deviceType = .nineInOne
                
            case .tv:
                deviceType = .tv
                
            case .dvd:
                deviceType = .dvd
                
            case .sat:
                deviceType = .sat
                
            case .appletv:
                deviceType = .appletv
                
            case .projector:
                deviceType = .projector
                
            case .dmx:
                
                let dmxGroupSettingController =
                    SHZoneDmxGroupSettingViewController()
                
                dmxGroupSettingController.currentZone =
                    currentZone
                
                navigationController?.pushViewController(
                    dmxGroupSettingController,
                    animated: true
                )
                
                return; // dmx 的方式也不同
            
            default:
                break
            }
            
        } else if selectController.isKind(of: SHZoneMoodViewController.self) {
            
            let editMoodController =
                SHZoneControlEditMoodViewController()
            
            editMoodController.currentZone = self.currentZone
            
            navigationController?.pushViewController(
                editMoodController,
                animated: true
            )
            
            return; // Mood是不同的方式
            
        } else if selectController.isKind(of: SHZoneLightViewController.self) {
            
            deviceType = .light
        
        } else if selectController.isKind(of: SHZoneShadeViewController.self) {
            
            deviceType = .shade
        
        } else if selectController.isKind(of: SHZoneFanViewController.self) {
            
            deviceType = .fan
        
        } else if selectController.isKind(of: SHZoneDryContactViewController.self) {
            
            deviceType = .dryContact
        
        } else if selectController.isKind(of: SHZoneTemperatureSensorViewController.self) {
            
            deviceType = .temperatureSensor
       
        } else if selectController.isKind(of: SHSceneControlViewController.self) {
            
            deviceType = .sceneControl
        
        } else if selectController.isKind(of: SHSequenceControlViewController.self) {
            
            deviceType = .sequenceControl
        
        } else if selectController.isKind(of: SHOtherControlViewController.self) {
            
            deviceType = .otherControl
        }
        
        let detailConroller =
            SHSystemDetailViewController()
        
        detailConroller.systemType = deviceType
        
        detailConroller.currentZone = currentZone
        
        navigationController?.pushViewController(
            detailConroller,
            animated: true
        )
        
    }
    
    /// 设置导航栏
    private func setupNavigationBar() {
        
        navigationItem.title = currentZone?.zoneName
        
        navigationItem.rightBarButtonItem =
            UIBarButtonItem(
                imageName: "setting",
                hightlightedImageName: "setting",
                addTarget: self,
                action: #selector(setArea),
                isLeft: false
        )
        
        let itemSize =
                UIDevice.is_iPad() ?
                (navigationBarHeight + statusBarHeight) :
                tabBarHeight
        
        let bounds =
            CGRect(x: 0,
                   y: 0,
                   width: itemSize,
                   height: itemSize
        )
        
        let leftItem =
            navigationItem.leftBarButtonItem?.customView
        
        let rightItem =
            navigationItem.rightBarButtonItem?.customView
        
        leftItem?.bounds = bounds
        rightItem?.bounds = bounds
    }
}


// MARK: - 子控制器 与 TabBar
extension SHAreaControlViewController {
    
    /// 设置tabBar
    private func setUpTabBar() {
        
        tabBar.backgroundImage =
            UIImage.resize("tabBarbackground")
        
        view.addSubview(tabBarScrollView)
        
        let count = children.count
        
        for i in 0 ..< count {
            
            let zoneControlButton = SHZoneControlButton()
            
            zoneControlButton.tag = i
            
            // 设置图片
            //  [zoneControlButton setBackgroundImage:[UIImage resizeImage:@"tabBarSelectedbackground"] forState:UIControlStateSelected];
            
            let normalImage =
                UIImage(named: deviceTypeNames[i] + "TabBar")
            
            zoneControlButton.setImage(normalImage,
                                       for: .normal
            )
            
            let selectedImage =
                UIImage(named: deviceTypeNames[i] +
                    "TabBar_highlighted"
            )
            
            zoneControlButton.setImage(selectedImage,
                                       for: .selected
            )
            
            // 获得按钮名称 -- 取消文字显示
            //        [zoneControlButton setTitle:[[SHLanguageTools shareLanguageTools] getTextFromPlist:@"MAIN_PAGE" withSubTitle:self.systemNames[i]] forState:UIControlStateNormal];
            
            tabBarScrollView.addSubview(zoneControlButton)
            
            zoneControlButton.addTarget(
                self,
                action: #selector(changeViewController(_:)),
                for: .touchUpInside
            )
            
            if i == 0 { // 默认第一个是选中状态
                
                changeViewController(zoneControlButton)
            }
        }
    }
        
    /// 选择不同的控制器
    @objc private func changeViewController(_ button: SHZoneControlButton) {
        
        if button == preivousButton {
            
            return
        }
        
        button.isSelected = !button.isSelected
        
        preivousButton?.isSelected =
            !(preivousButton?.isSelected ?? false)
        
        preivousButton = button
        
        if button.isSelected {
            
            selectedIndex = button.tag
        }
        
    }

    
    /// 添加子控制器
    private func addChildViewControllers() {
        
       deviceTypeIDs =
            SHSQLiteManager.shared.getSystemIDs(
                self.currentZone?.zoneID ?? 0
            )
        
        if deviceTypeIDs.isEmpty {
            
            SVProgressHUD.showInfo(withStatus:
                SHLanguageText.noData
            )
            
            return
        }
        
        let count = deviceTypeIDs.count
        
        for index in 0 ..< count {
            
            switch deviceTypeIDs[index] {
                
                // light
            case SHSystemDeviceType.light.rawValue:
                
                deviceTypeNames.append("alight")
                
                let lightController =
                    SHZoneLightViewController()
                
                lightController.currentZone = currentZone
                
                addChild(lightController)
                
                // hvac
            case SHSystemDeviceType.hvac.rawValue:
                
                deviceTypeNames.append("bhvac")
                
                let hvacController =
                    SHZoneDevicesViewController()
                
                hvacController.deviceType = .hvac
                
                hvacController.currentZone = currentZone
                
                addChild(hvacController)
                
            // audio
            case SHSystemDeviceType.audio.rawValue:
                
                deviceTypeNames.append("czaudio")
                
                let audioController =
                    SHZoneDevicesViewController()
                
                audioController.deviceType = .audio
                
                audioController.currentZone = currentZone
                
                addChild(audioController)
                
            // shade
            case SHSystemDeviceType.shade.rawValue:
                
                deviceTypeNames.append("dshades")
                
                let shadeController =
                    SHZoneShadeViewController()
                
                shadeController.currentZone = currentZone
                
                addChild(shadeController)
                
                // tv
            case SHSystemDeviceType.tv.rawValue:
                
                deviceTypeNames.append("etv")
                
                let tvController =
                    SHZoneDevicesViewController()
                
                tvController.deviceType = .tv
                
                tvController.currentZone = currentZone
                
                addChild(tvController)
                
            // sat
            case SHSystemDeviceType.sat.rawValue:
                
                deviceTypeNames.append("esat")
                
                let satController =
                    SHZoneDevicesViewController()
                
                satController.deviceType = .sat
                
                satController.currentZone = currentZone
                
                addChild(satController)
                
            // dvd
            case SHSystemDeviceType.dvd.rawValue:
                
                deviceTypeNames.append("edvd")
                
                let dvdController =
                    SHZoneDevicesViewController()
                
                dvdController.deviceType = .dvd
                
                dvdController.currentZone = currentZone
                
                addChild(dvdController)
                
            // appleTV
            case SHSystemDeviceType.appletv.rawValue:
                
                deviceTypeNames.append("eappletv")
                
                let appleTVController =
                    SHZoneDevicesViewController()
                
                appleTVController.deviceType = .appletv
                
                appleTVController.currentZone = currentZone
                
                addChild(appleTVController)
                
            // projector
            case SHSystemDeviceType.projector.rawValue:
                
                deviceTypeNames.append("eprojector")
                
                let projectorController =
                    SHZoneDevicesViewController()
                
                projectorController.deviceType = .projector
                
                projectorController.currentZone = currentZone
                
                addChild(projectorController)
                
            // mood
            case SHSystemDeviceType.mood.rawValue:
                
                deviceTypeNames.append("fmoods")
                
                let moodController =
                    SHZoneMoodViewController()
                
                moodController.currentZone = currentZone
                
                addChild(moodController)
                
            // fan
            case SHSystemDeviceType.fan.rawValue:
                
                deviceTypeNames.append("gfan")
                
                let fanController =
                    SHZoneFanViewController()
                
                fanController.currentZone = currentZone
                
                addChild(fanController)
                
            // floorHeating
            case SHSystemDeviceType.floorHeating.rawValue:
                
                deviceTypeNames.append("hFloorheating")
                
                let floorHeatingController =
                    SHZoneDevicesViewController()
                
                floorHeatingController.deviceType
                    = .floorHeating
                
                floorHeatingController.currentZone = currentZone
                
                addChild(floorHeatingController)
                
            // nineInOne
            case SHSystemDeviceType.nineInOne.rawValue:
                
                deviceTypeNames.append("jnieInOne")
                
                let nineInOneController =
                    SHZoneDevicesViewController()
                
                nineInOneController.deviceType
                    = .nineInOne
                
                nineInOneController.currentZone = currentZone
                
                addChild(nineInOneController)
                
                // 4Z/24Z
            case SHSystemDeviceType.dryContact.rawValue:
                
                deviceTypeNames.append("kdryInput")
                
                let nodeController =
                    SHZoneDryContactViewController()
                
                nodeController.currentZone = currentZone
                
                addChild(nodeController)
                
            // 4T
            case SHSystemDeviceType.temperatureSensor.rawValue:
                
                deviceTypeNames.append("ltemperature")
                
                let temperatureController =
                    SHZoneTemperatureSensorViewController()
                
                temperatureController.currentZone = currentZone
                
                addChild(temperatureController)
                
            // dmx
            case SHSystemDeviceType.dmx.rawValue:
                
                deviceTypeNames.append("mdmx")
                
                let dmxController =
                    SHZoneDevicesViewController()
                
                dmxController.deviceType = .dmx
                
                dmxController.currentZone = currentZone
                
                addChild(dmxController)
                
            // Scene
            case SHSystemDeviceType.sceneControl.rawValue:
                
                deviceTypeNames.append("nscene")
                
                let sceneController =
                    SHSceneControlViewController()
                
                sceneController.currentZone = currentZone
                
                addChild(sceneController)
                
            // Sequence
            case SHSystemDeviceType.sequenceControl.rawValue:
                
                deviceTypeNames.append("psequence")
                
                let sequenceController =
                    SHSequenceControlViewController()
                
                sequenceController.currentZone = currentZone
                
                addChild(sequenceController)
                
            // otherControl
            case SHSystemDeviceType.otherControl.rawValue:
                
                deviceTypeNames.append("qother")
                
                let otherController =
                    SHOtherControlViewController()
                
                otherController.currentZone = currentZone
                
                addChild(otherController)
                
            default:
                break
            }
        }
    }
}

// MARK: - 布局
extension SHAreaControlViewController {
    
    /// 布局子控件
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        tabBarScrollView.frame = tabBar.frame
        
        let itemHeight = tabBarHeight
        let itemWidth = tabBar.frame_width * 0.25 // (1/4)
        
        let startX =
            (tabBar.frame_width -
                CGFloat(children.count > 4 ? 4 : children.count) * itemWidth
                ) * 0.5
        
        let count = tabBarScrollView.subviews.count
        
        for i in 0 ..< count {
            
            let button = tabBarScrollView.subviews[i]
            
            button.frame =
                CGRect(x: CGFloat(button.tag) * itemWidth
                    + startX,
                       y: 0,
                       width: itemWidth,
                       height: itemHeight
            )
        }
        
        tabBarScrollView.contentSize =
            CGSize(width: itemWidth * CGFloat(children.count),
                   height: itemHeight
        )
    }
}
